Source/AcceleratorLattice/LatticeElements/Solenoid.H

```diff
@@ -0,0 +1,162 @@
+/* Copyright 2023 David Grote and David Bizzozero
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+#ifndef WARPX_ACCELERATORLATTICE_LATTICEELEMENTS_SOLENOID_H_
+#define WARPX_ACCELERATORLATTICE_LATTICEELEMENTS_SOLENOID_H_
+
+#include "LatticeElementBase.H"
+#include "HardEdged_K.H"
+
+#include <AMReX_REAL.H>
+#include <AMReX_ParmParse.H>
+#include <AMReX_GpuContainers.H>
+
+#include <string>
+#include <cmath>
+
+// Specifies a field that models the field generated by a solenoidal field,
+// an axisymmetric field expansion
+
+struct SolenoidDevice;
+
+struct Solenoid
+    : LatticeElementBase
+{
+
+    Solenoid ();
+
+    ~Solenoid () = default;
+
+    /**
+     * \brief Read in an element and add it to the lists
+     *
+     * @param[in] pp_element The ParmParse instance to read in the data
+     * @param[inout] z_location The current z location in the lattice
+     */
+    void
+    AddElement (amrex::ParmParse & pp_element, amrex::ParticleReal & z_location);
+
+    /**
+     * \brief Write the element information to the device
+     */
+    void
+    WriteToDevice ();
+
+    // Host variables read from input file
+    std::vector<amrex::ParticleReal> h_scale;
+    std::vector<amrex::ParticleReal> h_freq;
+    std::vector<amrex::ParticleReal> h_theta;
+    std::vector<std::vector<amrex::ParticleReal>> h_e_coef;
+    std::vector<std::vector<amrex::ParticleReal>> h_b_coef;
+
+    // Device variable read from input files
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_scale;
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_freq;
+    amrex::Gpu::DeviceVector<amrex::ParticleReal> d_theta;
+    amrex::Gpu::DeviceVector<std::vector<amrex::ParticleReal>> d_e_coef;
+    amrex::Gpu::DeviceVector<std::vector<amrex::ParticleReal>> d_b_coef;
```

Comment:
I don't think that this will work - `std::vector`` can't be used in device code. This might work
```
    amrex::Gpu::DeviceVector<amrex::ParticleReal*> d_e_coef;
    amrex::Gpu::DeviceVector<amrex::ParticleReal*> d_b_coef;
```
And changing how it is setup below.